src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -490,21 +476,28 @@ public CloseableIterator<PrimaryKeyWithSortKey> orderResultsBy(SSTableReader rea
         {
             // brute force using the in-memory compressed vectors to cut down the number of results returned
             var queryVector = vts.createFloatVector(orderer.vector);
-            return toMetaSortedIterator(this.orderByBruteForce(queryVector, rowIds, limit, rerankK), context);
+            return toMetaSortedIterator(this.orderByBruteForce(queryVector, segmentToOrdinalMap, limit, rerankK), context);
         }
+        // Create bits from the mapping
+        var bits = bitSetForSearch();
+        segmentToOrdinalMap.values().forEach(bits::set);
         // else ask the index to perform a search limited to the bits we created
         var queryVector = vts.createFloatVector(orderer.vector);
         var results = graph.search(queryVector, limit, rerankK, 0, bits, context, cost::updateStatistics);
         return toMetaSortedIterator(results, context);
     }
 
-    private Pair<SparseBits, IntArrayList> flatmapPrimaryKeysToBitsAndRows(List<PrimaryKey> keysInRange) throws IOException
+
+    /**
+     * Build a mapping of segment row id to ordinal for the given primary keys, skipping any that don't exist in this
+     * segment.
+     * @param keysInRange the primary keys to map
+     * @return a mapping of segment row id to ordinal
+     * @throws IOException
+     */
+    private Int2IntHashMap flatmapPrimaryKeysToBitsAndRows(List<PrimaryKey> keysInRange) throws IOException
     {
-        // if we are brute forcing the similarity search, we want to build a list of segment row ids,
-        // but if not, we want to build a bitset of ordinals corresponding to the rows.
-        // We won't know which path to take until we have an accurate key count.
-        var bits = bitSetForSearch();
-        IntArrayList rowIds = new IntArrayList();
+        var segmentToOrdinalMap = new Int2IntHashMap(keysInRange.size(), 0.65f, -1);
```

Comment:
nit: I'm a fan of using the "unofficial default" value of Integer.MIN_VALUE, it's more obvious if something goes wrong that it's not just an off-by-one somewhere

Reply:
Good to know, that makes sense to me.